style/js/shield.js

```diff
@@ -0,0 +1,384 @@
+        break;
+      }
+    }
+  } else {
+    shieldArtwork = shieldDef.backgroundImage;
+    textLayout = layoutShieldText(
+      c,
+      ctx,
+      ref,
+      shieldArtwork,
+      shieldDef.padding
+    );
+  }
+
+  //Special cases
+  if (ref.length == 0) {
+    shieldArtwork = ShieldDef.getNoRefArtwork(network);
+
+    if (shieldArtwork == null) {
+      return false;
+    }
+
+    c.width = shieldArtwork.data.width * spriteUpscale;
+    c.height = shieldArtwork.data.height * spriteUpscale;
+  }
+
+  loadShield(ctx, shieldArtwork);
+
+  if (shieldDef.notext != true) {
+    ctx.fillStyle = shieldDef.textColor;
+    drawShieldText(ctx, ref, textLayout);
+  }
+  /*
+  if (typeof shieldDef.modifiers !== "undefined") {
+    c = drawBanners(c, shieldDef.modifiers);
+  }
+  */
+  return true;
+}
+
+function drawShieldsToCanvas(c, ctx, network, ref) {
+  //Default draw size
+  c.width = 80;
+  c.height = 80;
+
+  switch (network) {
+    case "US:PA:Belt":
+      ctx.fillStyle = "white";
+      ctx.fillRect(0, 0, 80, 80);
+      ctx.lineWidth = 8;
+      ctx.strokeStyle = "black";
+      ctx.strokeRect(0, 0, 80, 80);
+
+      ctx.beginPath();
+      ctx.arc(40, 40, 22, 0, 2 * Math.PI, false);
+
+      switch (ref) {
+        case "Red Belt":
+          ctx.fillStyle = "#b01c2e";
+          break;
+        case "Orange Belt":
+          ctx.fillStyle = "#d97300";
+          break;
+        case "Yellow Belt":
+          ctx.fillStyle = "#f7d117";
+          break;
+        case "Green Belt":
+          ctx.fillStyle = "#006b54";
+          break;
+        case "Blue Belt":
+          ctx.fillStyle = "#003882";
+          break;
+        default:
+          return;
+      }
+
+      ctx.fill();
+
+      ctx.lineWidth = 4;
+      ctx.strokeStyle = "black";
+      ctx.stroke();
+      return true;
+
+    //Circle shields
+    case "US:DE":
+    case "US:IA":
+    case "US:MS":
+    case "US:NJ":
+    case "US:VA:Secondary":
+      if (ref == null || ref.length == 0) {
+        return false;
+      }
+
+      ctx.beginPath();
+      ctx.arc(40, 40, 37.5, 0, 2 * Math.PI, false);
+      ctx.fillStyle = "white";
+      ctx.fill();
+
+      ctx.lineWidth = 5;
+      ctx.strokeStyle = "black";
+      ctx.stroke();
+
+      ctx.fillStyle = "black";
+
+      var textLayout = layoutShieldText(c, ctx, ref, null, {
+        left: 11,
+        right: 11,
+        top: 11,
+        bottom: 11,
+      });
+      drawShieldText(ctx, ref, textLayout);
+
+      return true;
+  }
+  return false;
+}
+
+//Space between concurrent shields
+const spacer_size = 15;
+
+export function missingIconLoader(map, e) {
+  var id = e.id;
+
+  if (id == "shield_") {
+    return;
+  }
+
+  if (id == "spacer") {
+    map.addImage(
+      id,
+      {
+        width: spacer_size,
+        height: spacer_size,
+        data: new Uint8Array(4 * spacer_size * spacer_size),
+      },
+      {
+        pixelRatio: window.devicePixelRatio,
```

Comment:
This should actually be omitted unless you're going to multiply spacer_size by the DPR as well. The icon ends up half size as a result.